dvc/cache/base.py

```diff
@@ -101,40 +117,53 @@ def changed(self, path_info, hash_info):
         Args:
             path_info: dict with path information.
             hash: expected hash value for this data.
+            filter_info: an optional argument to target a specific path.
 
         Returns:
             bool: True if data has changed, False otherwise.
         """
 
+        hash_origin = path_info
+        path = filter_info or path_info
+
         logger.trace(
             "checking if '%s'('%s') has changed.", path_info, hash_info
         )
 
-        if not self.tree.exists(path_info):
-            logger.debug("'%s' doesn't exist.", path_info)
+        if not self.tree.exists(path):
+            logger.debug("'%s' doesn't exist.", path)
             return True
 
         if not hash_info:
             logger.debug("hash value for '%s' is missing.", path_info)
             return True
 
-        if self.changed_cache(hash_info):
+        if self.changed_cache(hash_info, filter_info=filter_info):
             logger.debug(
                 "cache for '%s'('%s') has changed.", path_info, hash_info
             )
             return True
 
-        actual = self.tree.get_hash(path_info)
+        # Instead of directly checking whether filter_info is present or not,
+        # we compare the target path with the path_info because sometimes
+        # filter_info might be equal to the path_info and in those cases we do
+        # not need to specialize the hash.
+        if path != path_info:
+            hash_origin, hash_info = self._get_specialized_hash_info(
+                path, hash_info, hash_origin
+            )
+
+        actual = self.tree.get_hash(hash_origin)
```

Comment:
The logic here is really confusing. Maybe just filter the hash and then use a helper method to check if the target has changed.

Reply:
I've kept the comparison logic in place, I think it seems a bit more simple now since we moved to the filtering part to the upper side. Could you check it again? (if you still find it complicated, I can move it out as well, as a separate function)